component/http/logging.go

```diff
@@ -0,0 +1,166 @@
+package http
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/beatlabs/patron/log"
+)
+
+var statusCodeLogger statusCodeLoggerHandler
+
+type intervalType uint32
+
+const (
+	included intervalType = iota // '[' or ']'
+	excluded                     // '(' or ')'
+)
+
+func init() {
+	cfg, _ := os.LookupEnv("PATRON_HTTP_STATUS_ERROR_LOGGING")
```

Comment:
Since we have an HTTP component and we are using a builder pattern i woner it it would be better to have this functionality as part of the http component.
The reasons are:
- Patron supports spinning up multiple HTTP componenets and with this setup both of them are not independent
- Having env vars so deep inside the code is something that we avoid in patron. Usually we resolve everything at the patron service level and pass the values in.

Reply:
Where would you see it being configured exactly?

Reply:
`createHTTPComponent()` in service.go. get the env var and use the builder to add this to the http component

Reply:
From what I can see, all the builder options are used while constructing the HTTP component (read timeout, write timeout, etc).
Yet, this logging parameter is used _at runtime_ in this function (which not a method with a state, just a pure stateless function): https://github.com/beatlabs/patron/blob/4bb0f6c654f7f77764b4535ec99aacd7affb0776/component/http/handler.go#L16
How would you map both worlds? It would require refactoring the handler.

Reply:
would making the handlers methods of a struct that is generated on setup work?

Reply:
Mmh there's a timeline problem I believe.

The handler is created in `NewRouteBuilder` which occurs before the creation of the HTTP component.
`createHTTPComponent` is called in `Run()` [here](https://github.com/beatlabs/patron/blob/6ad69ec15e2fec9bef595d443f79b9a8ecddb6f9/service.go#L439), the final step.

If we want to do the lookup in `createHTTPComponent` we would have to somehow _retrofit_ the HTTP handlers already created.

Perhaps a solution is to put the env var lookup in `NewRouteBuilder`?

Reply:
It just came to me. Check out middleware.go and the function `func logRequestResponse(corID string, w *responseWriter, r *http.Request) `. This is the lpace where to put our logging logic. Now we have to find a way to pass in the data for making the logging logic. What do you think?

Reply:
Okay but just for information, in a middleware that would require creating a custom `http.ResponseWriter` passed to the handler. Otherwise, it's not possible to retrieve the payload and status.

Also, where do you want the logic to parse the env var to be put?

Reply:
It exists already in line 140 in middleware.
the env var should be ideally live in patron service/ http component builder?

Reply:
Line 140 creates a `responseWriter` with status code = -1

Reply:
How about now,